File: examples/csharp/resource_generator/resource_generator.cs
/*
 * Copyright 2007-2024 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/**
 * @file resource_generator.cs
 *
 * @brief This example program provides a basic demonstration of using
 * the ResourceGenerator to publish resource messages onto the GMSEC
 * Bus.
 */

using GMSEC.API5;
using System;
using System.Text;


public class resource_generator
{
	public static int Main(string[] args)
	{
		// Set up connection configuration options, some from the command line,
		// others hard-coded.
		Config config = new Config(args);

		// Initialize log level for output
		InitializeLogLevel(config);

		// Validate messages before being sent to the bus
		config.AddValue("gmsec-msg-content-validate-send", "true");

		// Define the resource publish rate and how often resource are collected
		ushort publishRate     = 5;  // in seconds
		ushort sampleInterval  = 1;  // in seconds
		ushort averageInterval = 10; // in seconds

		try
		{
			// Create ResourceGenerator instance.
			using (ResourceGenerator rsrcgen = new ResourceGenerator(config, publishRate, sampleInterval, averageInterval, GetStandardFields()))
			{
				// Output information regarding the API version and connection
				Log.Info(Connection.GetAPIVersion());
				Log.Info("ResourceGenerator publish rate is: " + publishRate + " seconds");

				// Start the resource generator
				rsrcgen.Start();
				Log.Info("ResourceGenerator is running; use gmsub or other utility to monitor messages");

				Log.Info("Press <enter> to stop the ResourceGenerator");
				Console.ReadLine();

				// Stop the ResourceGenerator
				rsrcgen.Stop();
				Log.Info("ResourceGenerator has been stopped");
			}
		}
		catch (GmsecException e)
		{
			// If here, the most likely case is the middleware could not be
			// deduced, or a failure occurred when attempting to connect to
			// the GMSEC Bus.
			Log.Error("GmsecException: " + e.ToString());
			return -1;
		}

		return 0;
	}


	private static void InitializeLogLevel(Config config)
	{
		LogLevel level = Log.FromString( config.GetValue("loglevel", "info") );
		Log.SetReportingLevel(level);
	}


	private static FieldList GetStandardFields()
	{
		FieldList standardFields = new FieldList()
		{
			new StringField("MISSION-ID", "MY-MISSION", true),
			new StringField("SAT-ID-PHYSICAL", "MY-SAT-ID", true),
			new StringField("SAT-ID-LOGICAL", "MY-SAT-ID", true),
			new StringField("CONSTELLATION-ID", "MY-CNST", true),
			new StringField("FACILITY", "MY-FACILITY", true),
			new StringField("DOMAIN1", "MY-DOMAIN-1", true),
			new StringField("DOMAIN2", "MY-DOMAIN-2", true),
			new StringField("COMPONENT", "RSRC-GEN", true)
		};

		return standardFields;
	}
}

File: examples/c/resource_generator/resource_generator.c
/*
 * Copyright 2007-2025 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/**
 * @file resource_generator.c
 *
 * @brief This example program provides a basic demonstration of using
 * the ResourceGenerator to publish resource messages onto the GMSEC
 * Bus.
 */

#include "../utility.h"

#include <gmsec5_c.h>

#include <stdio.h>
#include <stdlib.h>


void checkStatus(const char* operation, GMSEC_Status status);

void getStandardFields(GMSEC_Field** standardFields, size_t* numFields);
void freeStandardFields(GMSEC_Field** standardFields, size_t numFields);


int main(int argc, char** argv)
{
	GMSEC_Status status = statusCreate();

	// Set up connection configuration options, some from the command line,
	// others hard-coded.
	GMSEC_Config config = configCreateWithArgs(argc, argv);

	// Initialize log level for output
	initializeLogLevel(config);

	// Validate messages before being sent to the bus
	configAddValue(config, "gmsec-msg-content-validate-send", "true", NULL);

	// Define the resource publish rate and how often resource are collected
	GMSEC_U16 publishRate     = 5;  // in seconds
	GMSEC_U16 sampleInterval  = 1;  // in seconds
	GMSEC_U16 averageInterval = 10; // in seconds

	// Acquire standard fields to be included with the resource messages
	GMSEC_Field* standardFields = {0};
	size_t       numFields = 0;

	getStandardFields(&standardFields, &numFields);


	// Create ResourceGenerator instance.
	GMSEC_ResourceGenerator rsrcgen = resourceGeneratorCreateWithFields(config,
	                                                                    publishRate,
	                                                                    sampleInterval,
	                                                                    averageInterval,
	                                                                    standardFields,
	                                                                    numFields,
	                                                                    status);
	checkStatus("resourceGeneratorCreateWithFields", status);

	// Release resources associated with standard fields
	freeStandardFields(&standardFields, numFields);

	// Output information regarding the API version and connection
    GMSEC_INFO(connectionGetAPIVersion());
    GMSEC_INFO("ResourceGenerator publish rate is: %u seconds", publishRate);

	// Start the resource generator
	resourceGeneratorStart(rsrcgen, status);
	checkStatus("resourceGeneratorStart", status);

	GMSEC_INFO("ResourceGenerator is running; use gmsub or other utility to monitor messages");

	GMSEC_INFO("Press <enter> to stop the ResourceGenerator");
	getchar();

	// Stop the ResourceGenerator
	resourceGeneratorStop(rsrcgen, NULL);

	GMSEC_INFO("ResourceGenerator has been stopped");

	// Cleanup resources
	resourceGeneratorDestroy(&rsrcgen);
	configDestroy(&config);
	statusDestroy(&status);

	return 0;
}


void checkStatus(const char* operation, GMSEC_Status status)
{
	if (statusHasError(status) == GMSEC_TRUE)
	{
		GMSEC_ERROR("%s failed; reason: %s", operation, statusGet(status));
		exit(-1);
	}
}


void getStandardFields(GMSEC_Field** standardFields, size_t* numFields)
{
	*numFields      = 8;
	*standardFields = malloc(*numFields * sizeof(GMSEC_Field));

	(*standardFields)[0] = stringFieldCreate("MISSION-ID", "MY-MISSION", GMSEC_TRUE, NULL);
	(*standardFields)[1] = stringFieldCreate("SAT-ID-PHYSICAL", "MY-SAT-ID", GMSEC_TRUE, NULL);
	(*standardFields)[2] = stringFieldCreate("SAT-ID-LOGICAL", "MY-SAT-ID", GMSEC_TRUE, NULL);
	(*standardFields)[3] = stringFieldCreate("CONSTELLATION-ID", "MY-CNST", GMSEC_TRUE, NULL);
	(*standardFields)[4] = stringFieldCreate("FACILITY", "MY-FACILITY", GMSEC_TRUE, NULL);
	(*standardFields)[5] = stringFieldCreate("DOMAIN1", "MY-DOMAIN-1", GMSEC_TRUE, NULL);
	(*standardFields)[6] = stringFieldCreate("DOMAIN2", "MY-DOMAIN-2", GMSEC_TRUE, NULL);
	(*standardFields)[7] = stringFieldCreate("COMPONENT", "RSRC-GEN", GMSEC_TRUE, NULL);
}


void freeStandardFields(GMSEC_Field** standardFields, size_t numFields)
{
	size_t i;

	for (i = 0; i < numFields; ++i)
	{
		fieldDestroy(&(*standardFields)[i]);
	}

	free(*standardFields);

	*standardFields = NULL;
}
